### PR TITLE
feat: reworked varnish_cache plugin

### LINKED ISSUE #5622: varnish plugin should cleanup backend metric name
## Feature Request
Varnish plugin dumps the backend list as metric names, making this fields mostly useless for query. We need a common name for the metric and the backend names and profiles should be pushed to labels

### Proposal:
Either parse the varnish VBE metrics and extract the labels for the backends instead of dumping then in the metric name OR rename varnish VBE metric to a more friendly name like varnish_vbe_* , so we can try to fix this outside by renaming the metrics (at least prometheus can do this, do not know about other tools)

### Current behavior:
right now if we request the varnish plugin to gather backend metrics, we get this:

image

I can't even try to rename/relabel this in prometheus, as it's directly a varnish_*, so i can't create a rule to parse this as it would also parse all other metrics. There is no common anchor for all backend metrics names, in my example, the boot is the profile name, loading more profiles i can get any name (eg: profile named cache), so it can collide with existent metrics

### Desired behavior:
The current setup is good for all metrics except VBE, with it, its mostly useless, i can't use those names in anything as it would require me to know the profile, backend name and ip. instead it should take the VBE metric as a special metric and extract the labels from the name and create the varnish_vbe_{comm,happy,req,etc}

### Use case: [Why is this important (helps with prioritizing requests)]
Getting backend metrics and health status in varnish is important information, but the plugin takes the VBE metric as a normal metric and it should be special as it can have random info. Current setup can overload any system with useless metric names, making everything slow

### PR DESCRIPTION
**Pull request by rhajek**
Resolves #5622 #6832

Added reworked version of the VarnishCache input plugin based on json parsing of `varnishstat -j`.
Conversion to metrics from Varnish stats is reworked, it solves issue with high cardinality when reloading Varnish server, parses  backends into tags, supports "VBE.*" metrics.

- [x] Updated associated README.md.
- [x] Wrote appropriate unit tests.

### CODE REVIEW

File: plugins/inputs/varnish_cache/varnish_cache.go
Comment (ismaelpuerto) [inline plugins/inputs/varnish_cache/varnish_cache.go:290]: This approach only work if you reload services, but if you load vcl manually with custom name doesn't valid

Comment (rhajek) [reply plugins/inputs/varnish_cache/varnish_cache.go]: I reworked this PR, now it is part of the varnish plugin, varnishadm tool is used to get active vcl.

File: plugins/inputs/varnish/varnish_test.go
Comment (helenosheaa) [inline plugins/inputs/varnish/varnish_test.go:582]: Are these flags and formats something that is provided from varnish metrics?

File: plugins/inputs/varnish/varnish.go
Comment (helenosheaa) [inline plugins/inputs/varnish/varnish.go:198]: Is this missing a return nil, if there's no error?

File: plugins/inputs/varnish/varnish.go
Comment (rhajek) [reply plugins/inputs/varnish/varnish.go:369]: I suggest to support varnish from version 6.0.2, which starts to support json output `varnishadm vcl.list -j` - https://github.com/varnishcache/varnish-cache/pull/2783

Is ok, that we drop support for v2 metrics for varnish older than 6.0.2 ?

File: plugins/inputs/varnish/varnish.go
Comment (rhajek) [reply plugins/inputs/varnish/varnish.go:249]: My goal was also to fix https://github.com/influxdata/EAR/issues/2289 and strip the active VCL name from field name for v1 metrics. If we want v1 to stay fully backward compatible, we should revert this and remove also unnecessary  `varnishadm vcl.list` check.

File: plugins/inputs/varnish/varnish.go
Comment (ismaelpuerto) [reply plugins/inputs/varnish/varnish.go:369]: Hello @helenosheaa this is the steps to load manually VCL with `varnishadm` command

```
[root@server ~]# varnishadm vcl.list
active      auto/warm         16 boot

[root@server ~]# varnishadm vcl.load myvclfile /etc/varnish/default2.vcl 
VCL compiled.

[root@server ~]# varnishadm vcl.list
active      auto/warm         32 boot
available   auto/warm          0 myvclfile

[root@server ~]# varnishadm vcl.use myvclfile
VCL 'myvclfile' now active

[root@server ~]# varnishadm vcl.list
available   auto/warm          0 boot
active      auto/warm         16 myvclfile
```

Comment (rhajek) [reply plugins/inputs/varnish/varnish.go]: I reworked parsing active vcl to use JSON from `varnishadm vcl.list -j`

File: plugins/inputs/varnish/varnish.go
Comment (rhajek) [reply plugins/inputs/varnish/varnish.go:395]: According https://varnish-cache.org/docs/6.5/whats-new/upgrading-6.5.html
Varnish 6.5 introduces the change in json structure which adds `counters` element:

Before
```
{
  "timestamp": "YYYY-mm-ddTHH:MM:SS",
  "MGT.uptime": {
    ...
  },
  ...
}
```

After 6.5:
```
{
  "version": 1,
  "timestamp": "YYYY-mm-ddTHH:MM:SS",
  "counters": {
    "MGT.uptime": {
      ...
    },
    ...
  }
}
```

File: plugins/inputs/varnish/varnish.go
Comment (anthosz) [reply plugins/inputs/varnish/varnish.go]: Hi,

The issue to keep nonactive VCL's is that it can generate huge of metrics that seems useless.

It can be very good to skip all inactive vcls.

There is a similar PR available on https://github.com/influxdata/telegraf/pull/6832 (but seems abandoned) related to https://github.com/influxdata/telegraf/issues/5622

File: plugins/inputs/varnish/varnish_test.go
Comment (rhajek) [reply plugins/inputs/varnish/varnish_test.go:573]: Flags and formats are used to display counter in varnishstat interactive text UI. `flag="b"` is used only for `happy` counter, which contains health probes encoded into btimap. Happy probes are bits set to 1, and the unhappy ones are set to 0. The highest bits represent the oldest probes.

File: plugins/inputs/varnish/varnish.go
Comment (rhajek) [reply plugins/inputs/varnish/varnish.go:52]: instance_name is optional, there is no needed to have special argument in runner, ["-n", "instance"] is appended in  prepareCmdArgs() for both `varnishstat` and `varnishadm` tools.

File: plugins/inputs/varnish/varnish.go
Comment (rhajek) [reply plugins/inputs/varnish/varnish.go:282]: By default json decoder serialize number into floats.  `varnishstat` produces integer counters, and we also want to store them as integers. So `dec.UseNumber()` force json decoder to use json.Number which supports parsing of long uint64 numbers.

Comment (rhajek) [reply plugins/inputs/varnish/varnish.go:282]: test failure in circle-ci was related to unordered json elements in map

File: plugins/inputs/varnish_cache/varnish_cache.go
Comment (helenosheaa) [reply plugins/inputs/varnish_cache/varnish_cache.go]: @ismaelpuerto is this concern resolved?

File: plugins/inputs/varnish/varnish.go
Comment (helenosheaa) [reply plugins/inputs/varnish/varnish.go]: I'm trying to clarify whether this approach will work in either case if the return from varnishadm list is:

varnishadm vcl.list
`active      auto/warm         16 boot `

or:
varnishadm vcl.list
`active auto warm 0 boot`

I think it's ok to only support varnish 6.0.2, but it needs to be part of the metric version 2 change and be included in the documentation.

File: plugins/inputs/varnish/varnish.go
Comment (helenosheaa) [reply plugins/inputs/varnish/varnish.go]: I understand but we try to keep backwards compatibility as up until now nonactive vcl's were reported. The approach is to give users the option to opt in to new behavior rather than changing behavior on an upgrade of Telegraf.

File: plugins/inputs/varnish/README.md
Comment (helenosheaa) [inline plugins/inputs/varnish/README.md:24]: Can we add something about the version of varnish which is supported by metric_version=2?

File: plugins/inputs/varnish/varnish.go
Comment (rhajek) [reply plugins/inputs/varnish/varnish.go]: I added requirement for varnish 6.0.2+ to README.md.

output from `varnishadm vcl.list -j` looks like
```json
[ 2, ["vcl.list", "-j"], 1631691282.104,
  {
    "status": "active",
    "state": "auto",
    "temperature": "warm",
    "busy": 0,
    "name": "boot"
}
]
```
so parsing `"name": "boot"` should not be problematic.

File: plugins/inputs/varnish/README.md
Comment (rhajek) [reply plugins/inputs/varnish/README.md]: Added varnish 6.0.2+ requirement into telegraf config example in README

File: plugins/inputs/varnish/varnish.go
Comment (helenosheaa) [reply plugins/inputs/varnish/varnish.go]: ok thanks for clarifying

File: plugins/inputs/varnish/varnish.go
Comment (powersj) [inline plugins/inputs/varnish/varnish.go:428]: This is where `id` is getting added as a tag and why I am asking at a high level why we are increasing the cardinality on that value. Is that something users want to index against?

It seems concerning to add anything that doesn't match `_vcl` or `_field` as a tag as any change to the data format is going to cause cardinality to increase.

**Review by powersj — COMMENTED** [review]
Thanks for this MP. A couple of small in-line comments and questions and one higher-level question:

Can you explain why [version 2 metrics](https://paste.ubuntu.com/p/scGJzs5syG/) include the `id` tag while [version 1 metrics](https://paste.ubuntu.com/p/WsnPcSK2w5/) omit it? This seems to greatly increase the cardinality even on my little system running varnish + apache2 with the default configs. See my comment in the `parseMetricV2` function.

Given one purpose of this MP is to reduce cardinality this seems to do the opposite as I have run it above.

Thanks!

edit: updated link to version 2 metrics

File: plugins/inputs/varnish/varnish.go
Comment (rhajek) [reply plugins/inputs/varnish/varnish.go:428]: Thanks for review. Id tag is added by the last default regexp rule (?P<id>)

```go
//generic metric like MSE_STORE.store-1-1.g_aio_running_bytes_write
regexp.MustCompile(`([\w\-]*)\.(?P<id>[\w\-.]*)\.([\w\-]*)`),
```

My idea was not to increase the cardinality, but make the field names corresponding with https://varnish-cache.org/docs/trunk/reference/varnish-counters.html

Cardinality issue was caused because varnish adds timestamps into its metrics names when reloading. This should be solved by tripping vcl name from the name. 

Example: we have following varnish metrics
```
  "MEMPOOL.busyobj.allocs"
  "MEMPOOL.ssl_buf.allocs"
  "MEMPOOL.req0.allocs"
  "MEMPOOL.sess0.allocs"
  "MEMPOOL.req1.allocs"
  "MEMPOOL.sess1.allocs"
```

Middle part identifies the name of resource. There are limited number of these so it is suitable for tagging/indexing.
We reduce number of tags but increase number of fields. I think cardinality in this case is the same. 

varnish,section=MEMPOOL,id=busyobj alloc=xxxx
varnish,section=MEMPOOL,id=ssl_buf alloc=xxxx
varnish,section=MEMPOOL,id=req0 alloc=xxxx

varnish,section=MEMPOOL busyobj.alloc=xxxx
varnish,section=MEMPOOL ssl_buf.alloc=xxxx
varnish,section=MEMPOOL req0.alloc=xxxx

Biggest disadvantage of this approach i see is that this approach will not be backward compatible with v1.
Varnish metrics can have various names, it tried to cover that in 'plugins/inputs/varnish/varnish_test.go:441' 

@helenosheaa @ismaelpuerto @anthosz  So should we remove the id tag and add middle part into field name?

File: plugins/inputs/varnish/varnish.go
Comment (rhajek) [reply plugins/inputs/varnish/varnish.go:368]: The only documentation that i found was:
https://varnish-cache.org/docs/6.6/reference/varnish-cli.html#json
and https://varnish-cache.org/docs/6.6/reference/varnish-cli.html#vcl-list-j

version "2" is still the same from 6.0.3  released in 9 Feb 2019

May be, check for the presence of "vcl.list", "-j" is enough and we can ignore version.

Unmarshal json array like this 

```json
[ 2, ["vcl.list", "-j"], 1631878235.913,
  {
    "status": "active",
    "state": "auto",
    "temperature": "warm",
    "busy": 0,
    "name": "boot"
}
]
```
into object is definitely better, I will try to fix this.

File: plugins/inputs/varnish/varnish.go
Comment (helenosheaa) [reply plugins/inputs/varnish/varnish.go:422]: If there's a static number of values that will be set as id=tags I don't think it will be as much of cardinality problem as we were seeing with vcl name where fields included timestamps or ip's so they were constantly increasing. 

However what is the advantage of splitting it out into a tag? The more consistency between v1 and v2 unless there is a value add I think the better.

That said I believe the build from this PR is being tested so shall we make a decision on this once we get feedback from that?

Comment (rhajek) [reply plugins/inputs/varnish/varnish.go:422]: Splitting varnish metric names into tags can be useful when creating dashboards. Represent a newly added resource as a tag looks more natural to me, it will be automatically shown in the dashboard if you use generic queries and filters. 

But, may be, more important is to keep v2 as much as possible backward compatible and not break existing things. 
User can always use custom regexps in telegraf configuration to create custom tag splitting if needed.

**Comment by helenosheaa** [comment]
hey @ismaelpuerto would you mind testing this build with the two different metric versions?

**Comment by ismaelpuerto** [comment]
Hello, I just test with both version and It looks fine, For this test I used varnish-plus-6.0.8r4 and I tested metrics like kvstore, vha and [accounting](https://docs.varnish-software.com/varnish-cache-plus/vmods/accounting/), regarding this metric, maybe it make sense that exists a tag with value of \<namespace\>, but I can understand that this a special case with a low audience and right now, it's not priority.

Regarding tags, for me is fine have "section" only, I don't need the name of vcl as tag because we have metric "n_vcl" to know if there was loaded a new vcl.

**Comment by rhajek** [comment]
> Hello, I just test with both version and It looks fine, For this test I used varnish-plus-6.0.8r4 and I tested metrics like kvstore, vha and [accounting](https://docs.varnish-software.com/varnish-cache-plus/vmods/accounting/), regarding this metric, maybe it make sense that exists a tag with value of <namespace>, but I can understand that this a special case with a low audience and right now, it's not priority.
> 
> Regarding tags, for me is fine have "section" only, I don't need the name of vcl as tag because we have metric "n_vcl" to know if there was loaded a new vcl.

To add a namespace tag for accounting metrics you can add following regexp into telegraf config. Should I include this rule to the defaults ?

[[inputs.varnish]]
    regexps = ['^ACCG\.(?P<namespace>[\w\-]*)\.(?P<_field>[\w\-\.]*)']

**Comment by helenosheaa** [comment]
I think instead of adding it to defaults as it's potentially low audience right now, just add it as an example in the README so people can add if they want to.

**Comment by ivankudibal** [comment]
@popey - the approver is unclear

**Comment by dg-nvm** [comment]
@helenosheaa can we get artifacts rebuild again ? Without this I cannot test changes

**Comment by powersj** [comment]
@rhajek can you rebase on master and push, please? Artifacts are only available for 30-days and there are some folks who want to test this PR out. Updating the PR with a rebase should trigger the tests and package builds again.

If users feel comfortable building the go binary themselves they can also do the following:

```
git clone https://github.com/rhajek/telegraf -b varnish_cache
cd telegraf
make
```

The above will produce a binary for use on the system it was run on (i.e. if you build on `linux/amd64`, it will be a `linux/amd64` binary).

**Comment by dg-nvm** [comment]
@powersj I had to upgrade Golang to do it but it worked and I will use my compiled one to test it

**Comment by dg-nvm** [comment]
@rhajek I still see plenty of `reload_20211108_110316_28415` fields on new configuration:

varnishstat (varnish-6.0.6 revision 29a1a8243dbef3d973aec28dc90403188c1dc8e7)
Telegraf unknown (git: varnish_cache 40c47357)

```
[[inputs.varnish]]
  #use_sudo = false
  #binary = "/bin/varnishstat"
  stats = ["all"]
  #name = instanceName
#  fielddrop = ["reload_*"]
```

Do we have to add some special configuration to get rid of these multiple fields ?

image

**Comment by rhajek** [comment]
> @rhajek I still see plenty of `reload_20211108_110316_28415` fields on new configuration:

@dg-nvm thanks for testing this PR, you have to use `metric_version = 2` in your telegraf.config to enable nonactive vcls removal.

**Comment by anthosz** [comment]
Hi @rhajek ,

Thank you very much for the works, I just tested with the artifacts of this day and seem's to be good: I have only the active vcls.

In addition, I have one remark concerning the "unhealthy" field, I cannot find it via varnishadm and it's always set to zero (I tried to stop the backend -> instance sick on varnish side -> but unhealthy value is always set to 0)

**Comment by dg-nvm** [comment]
@anthosz @rhajek I checked that too and yes, when using instance name command line looks like this:

`/bin/varnishadm vcl.list -j -n /instance-name`

Instead of:

`/bin/varnishadm -n /instance-name vcl.list -j`

Order of arguments matters, command and -j have to be at the end

**Comment by rhajek** [comment]
I fixed how to setup a custom instance name when using custom arguments in README.md.

**Comment by powersj** [comment]
@rhajek can you do a final rebase on master and resolve the conflicts on the readme? We can do a final review and land this if it looks good.

**Comment by rhajek** [comment]
> @rhajek can you do a final rebase on master and resolve the conflicts on the readme? We can do a final review and land this if it looks good.

@powersj PR is rebased and ready for approval.

**Review by sspaink — APPROVED** [review]

**Comment by anthosz** [comment]
Dear @sspaink ,

Do you know when it will be released?

Best regards,

**Comment by anthosz** [comment]
Hi @sspaink ,

I don't see any new release since the 16th (I prefer to use a "stable" version). Do you have an estimation?

Best regards,

**Comment by powersj** [comment]
> Do you have an estimation?

Plan is to start the release tomorrow - we decided to hold off a little bit so we could land some other snmp changes.

**Comment by sspaink** [comment]
@anthosz v1.22.0 has been released: https://github.com/influxdata/telegraf/releases/tag/v1.22.0 !

**Comment by anthosz** [comment]
@sspaink  thank you! \o/

**Comment by anthosz** [comment]
Good job to all for this PR, all works well :)

**Pull request merged**
